gatsby-node.js

```diff
@@ -194,60 +84,179 @@ async function doQuery({
-            'when enablePartialUpdates is true, the objects must have at least one of the match fields. Current object:\n' +
-              JSON.stringify(curObj, null, 2) +
-              '\n' +
-              'expected one of these fields:\n' +
-              matchFields.join('\n')
-          );
-        }
+function groupQueriesByIndex(queries = [], config) {
+  const { indexName: mainIndexName } = config;
 
-        const ID = curObj.objectID;
-        let extObj = algoliaObjects[ID];
+  return queries.reduce((groupedQueries, queryOptions) => {
+    const { indexName = mainIndexName } = queryOptions;
 
-        /* The object exists so we don't need to remove it from Algolia */
-        delete algoliaObjects[ID];
-        delete toRemove[ID];
+    return {
+      ...groupedQueries,
+      [indexName]: [
+        ...(groupedQueries.hasOwnProperty(indexName)
+          ? groupedQueries[indexName]
+          : []),
+        queryOptions,
+      ],
+    };
+  }, {});
+}
 
-        if (!extObj) return true;
+/**
+ * Run all queries for a given index, then make any updates / removals necessary
+ */
+async function runIndexQueries(
+  indexName,
+  queries = [],
+  { client, activity, graphql, config }
+) {
+  const {
+    settings: mainSettings,
+    chunkSize = 1000,
+    enablePartialUpdates = false,
+    matchFields: mainMatchFields = ['modified'],
+  } = config;
 
-        return matchFields.some(field => extObj[field] !== curObj[field]);
-      });
+  setStatus(
+    activity,
+    `Running ${queries.length} ${
+      queries.length === 1 ? 'query' : 'queries'
+    } for index ${indexName}...`
+  );
 
-      Object.keys(algoliaObjects).forEach(objectID => {
-        // if the object has one of the matchFields, it should be removed,
-        // but objects without matchFields are considered "not controlled"
-        // and stay in the index
-        if (matchFields.some(field => algoliaObjects[objectID][field])) {
-          toRemove[objectID] = true;
-        }
-      });
-    }
+  const objectMapsByQuery = await Promise.all(
+    queries.map(query => getObjectsMapByQuery(query, graphql))
+  );
 
-    setQueryStatus(
-      `Partial updates – [insert/update: ${hasChanged.length}, total: ${objects.length}]`
+  const allObjectsMap = objectMapsByQuery.reduce((acc, objectsMap = {}) => {
+    return {
+      ...acc,
+      ...objectsMap,
+    };
+  }, {});
+
+  setStatus(
+    activity,
+    `${queries.length === 1 ? 'Query' : 'Queries'} resulted in a total of ${
+      Object.keys(allObjectsMap).length
+    } results`
+  );
+
+  const index = client.initIndex(indexName);
+  const tempIndex = client.initIndex(`${indexName}_tmp`);
+  const indexToUse = await getIndexToUse({
+    index,
+    tempIndex,
+    enablePartialUpdates,
+  });
+
+  let toIndex = {}; // used to track objects that should be added / updated
+  const toRemove = {}; // used to track objects that are stale and should be removed
+
+  // let objectsToIndex = [];
```

Comment:
there's some commented code that can be removed

Reply:
Nice catch, will remove